trailblazer/apps/slurm/api.py

```diff
@@ -0,0 +1,62 @@
+from typing import List, Dict, Optional
+
+from trailblazer.apps.slurm.models import SqueueResult
+from trailblazer.constants import FileFormat, SlurmJobStatus, TrailblazerStatus
+from trailblazer.exc import EmptySqueueError
+from trailblazer.io.controller import ReadStream
+from trailblazer.apps.slurm.utils import formatters
+
+
+def get_squeue_result(squeue_response: str) -> SqueueResult:
+    """Return SqueueResult object from squeue response.
+    Raises:
+        TrailblazerError: when no entries were returned by squeue command.
+    """
+    if not squeue_response:
+        raise EmptySqueueError("No jobs found in SLURM registry")
+    squeue_response_content: List[dict] = ReadStream.get_content_from_stream(
+        file_format=FileFormat.CSV,
+        stream=squeue_response,
+        read_to_dict=True,
+    )
+    return SqueueResult(jobs=squeue_response_content)
+
+
+def reformat_squeue_result_job_step(data_analysis: str, job_step: str) -> str:
+    """Standardise job step string according to data anlaysis."""
+    formatter_func = formatters.formatter_map.get(data_analysis, formatters.reformat_undefined)
+    return formatter_func(job_step=job_step)
+
+
+def get_current_analysis_status(jobs_status_distribution: dict) -> Optional[str]:
+    """Return current analysis status based on jobs status distribution."""
+    is_ongoing: bool = bool(
+        jobs_status_distribution.get(SlurmJobStatus.RUNNING)
+        or jobs_status_distribution.get(SlurmJobStatus.PENDING)
+    )
+
+    analysis_status_map: Dict[str, Optional[str]] = {
+        SlurmJobStatus.FAILED: TrailblazerStatus.ERROR if is_ongoing else TrailblazerStatus.FAILED,
+        SlurmJobStatus.TIME_OUT: TrailblazerStatus.ERROR
+        if is_ongoing
+        else TrailblazerStatus.FAILED,
+        SlurmJobStatus.COMPLETED: TrailblazerStatus.COMPLETED
+        if jobs_status_distribution.get(SlurmJobStatus.COMPLETED) == 1
+        else None,
+        SlurmJobStatus.PENDING: TrailblazerStatus.PENDING
+        if jobs_status_distribution.get(SlurmJobStatus.PENDING) == 1
+        else None,
+        SlurmJobStatus.RUNNING: TrailblazerStatus.RUNNING,
+        SlurmJobStatus.CANCELLED: None if is_ongoing else TrailblazerStatus.CANCELLED,
+    }
+
+    for status in [
+        SlurmJobStatus.FAILED,
+        SlurmJobStatus.TIME_OUT,
+        SlurmJobStatus.COMPLETED,
+        SlurmJobStatus.PENDING,
+        SlurmJobStatus.RUNNING,
+        SlurmJobStatus.CANCELLED,
```

Comment:
This ordering is important right? If so this feels a bit sketchy. Will try to find an alternative tomorrow

Reply:
How about having status as a property for `SqueueResult`? Maybe something like this could be more clear:
```python
class SqueueResult(BaseModel):
    """Model used to parse SLURM squeue output."""

    jobs: List[SqueueJob]
    jobs_status_distribution: Optional[Dict[str, float]]

    @validator("jobs_status_distribution", always=True)
    def set_jobs_status_distribution(
        cls, _: Optional[List[SqueueJob]], values: Dict[str, Any]
    ) -> Optional[Dict[str, float]]:
        """Set the job status distribution."""
        jobs: List[SqueueJob] = values["jobs"]
        status_distribution: Dict[str, float] = {}
        for job in jobs:
            status_distribution[job.status] = status_distribution.get(job.status, 0) + 1
        for status in status_distribution:
            status_distribution[status] = round(status_distribution[status] / len(jobs), 2)
            return status_distribution

    @property
    def is_ongoing(self):
        return bool(
            self.jobs_status_distribution.get(SlurmJobStatus.RUNNING)
            or self.jobs_status_distribution.get(SlurmJobStatus.PENDING)
        )

    @property
    def only_one_status(self):
        return len(self.jobs_status_distribution) == 1

    @property
    def status(self):
        """Aggregates the statuses of all jobs into a status for a whole case. If multiple statuses
        the order is decided by STATUS_PRIORITY_LIST."""
        if self.only_one_status:
            only_status: str = self.jobs_status_distribution.popitem()[0]
            return SlurmTrailblazerStatusMap[only_status]

        aggregation_status = SlurmJobStatus.FAILED
        for status in STATUS_PRIORITY_LIST:
            if status in self.jobs_status_distribution.keys():
                aggregation_status = status
                break

        if aggregation_status == SlurmJobStatus.FAILED and self.is_ongoing:
            return TrailblazerStatus.ERROR
        return SlurmTrailblazerStatusMap[aggregation_status]


STATUS_PRIORITY_LIST = [
    SlurmJobStatus.FAILED,
    SlurmJobStatus.TIME_OUT,
    SlurmJobStatus.CANCELLED,
    SlurmJobStatus.RUNNING,
]

FAILED_STATUSES = [SlurmJobStatus.FAILED, SlurmJobStatus.TIME_OUT]

SlurmTrailblazerStatusMap: Dict[str, str] = {
    SlurmJobStatus.FAILED: TrailblazerStatus.FAILED,
    SlurmJobStatus.TIME_OUT: TrailblazerStatus.FAILED,
    SlurmJobStatus.COMPLETED: TrailblazerStatus.COMPLETED,
    SlurmJobStatus.PENDING: TrailblazerStatus.PENDING,
    SlurmJobStatus.RUNNING: TrailblazerStatus.RUNNING,
    SlurmJobStatus.CANCELLED: TrailblazerStatus.CANCELLED,
}

```

Reply:
@ChrOertlin  @Vince-janv I made a new attempt at this based on your suggestions. Please have a look.

Reply:
I think it looks much better 👍